Notepad.Extensions.Logging/NotepadLoggerProvider.cs

```diff
@@ -1,25 +1,46 @@
-﻿using System.Text;
+﻿using System;
+using System.Text;
 using Microsoft.Extensions.Logging;
 using Microsoft.Extensions.ObjectPool;
+using Microsoft.Extensions.Options;
 
 namespace Notepad.Extensions.Logging
 {
+    [ProviderAlias("Notepad")]
```

Comment:
I'm starting to wonder if `Notepad` and `Notepad++` should be two separate loggers (though with shared code, of course)

Reply:
Different `ILogger` classes or `ILoggerProvider`? I think it should be single provider, that can be configured to different applications kind. So it will writes only in one application.

Reply:
I think it should be very unlikely that someone would want to log to _both_ because that's what two loggers would imply. I think it would be better to just add support for it in one logger.

However regarding Notepad++, since this is a more common tool for developers to use (compared to notepad), maybe logging to it automatically this would be a bit troublesome and distracting. So maybe we should look for a more explicit opt-in mechanism. For example requiring users to name the window in a specific way.

Reply:
Some naming format will be realy helpfull for Notepad++. It will allow us to change regex on something like `logging(?:-(\S+))? - Notepad\+\+` and if user provide name to search it should be equal to the first group value.

Reply:
From a developer experience perspective I think it would be easier for users to call:

```csharp
AddNotepad(o => o.DocumentName = "MyDocumentName");
```

or:

```csharp
AddNotepadPlusPlus(o => o.DocumentName = "MyDocumentName")
```

and then for us to automatically match it to the window title (e.g. `*MyDocumentName - Notepad`).

And if they don't specify a document name, then we look for the defaults - `Untitled` for Notepad, and `/new \d+/` for NP++.

For now its probably simpler to have one logger that supports both (i.e. just `AddNotepad`), though.

Reply:
It can be solved by adding internal property to `NotepadLoggerOptions` and calling configure method.
```csharp
public static ILoggingBuilder AddNotepad(this ILoggingBuilder builder)
    => AddProvider(builder, WindowKind.Notepad);

public static ILoggingBuilder AddNotepad(this ILoggingBuilder builder, Action<NotepadLoggerOptions> configure)
    => AddProvider(builder, WindowKind.Notepad, configure);

public static ILoggingBuilder AddNotepadPlusPlus(this ILoggingBuilder builder)
    => AddProvider(builder, WindowKind.NotepadPlusPlus);

public static ILoggingBuilder AddNotepadPlusPlus(this ILoggingBuilder builder, Action<NotepadLoggerOptions> configure)
    => AddProvider(builder, WindowKind.NotepadPlusPlus, configure);

public static ILoggingBuilder AddProvider(this ILoggingBuilder builder, WindowKind kind, Action<NotepadLoggerOptions> configure)
{
    if (configure == null)
    {
        throw new ArgumentNullException(nameof(configure));
    }
    AddProvider(builder, kind);
    builder.Services.Configure(configure);
    return builder;
}

static ILoggingBuilder AddProvider(ILoggingBuilder builder, WindowKind kind)
{
    builder.AddConfiguration();

    builder.Services.TryAddEnumerable(ServiceDescriptor.Singleton<ILoggerProvider, NotepadLoggerProvider>());
    LoggerProviderOptions.RegisterProviderOptions<NotepadLoggerOptions, NotepadLoggerProvider>(builder.Services);

    if (kind != default)
    {
        builder.Services.Configure<NotepadLoggerOptions>(opt => opt.Application = kind);
    }

    return builder;
}
```

And later use this property for search like window name. What do you think? Of course it can be done after this PR as next enhancment.

Reply:
> From a developer experience perspective I think it would be easier for users to call

In this way we will unable to support multilanguage. In some cultures windows uses dash character instead of minus (`"–".Equals("-")` is false). So provide only document name will break multilanguage support for notepad. Notepad++ will work well.

Reply:
Good thinking. I'm probably gonna have to build up a list then of how Notepad[++] builds its title bar in every language... 😞